File: Sources/Domain/ConnectionDeltaTracker.swift
struct ConnectionDeltaTracker: Sendable {
  private var previousKernelTotal: TrafficBytes?
  private var previousConnections: [String: TrafficBytes] = [:]

  mutating func consume(
    _ snapshot: ConnectionTrafficSnapshot,
    classifier: ProxyClassifier
  ) -> ConnectionDeltaResult {
    guard let previousKernelTotal else {
      establishBaseline(from: snapshot)
      return .baselineEstablished
    }

    guard
      let kernelDelta = TrafficBytes.nonnegativeDelta(
        current: snapshot.kernelTotal,
        previous: previousKernelTotal
      )
    else {
      establishBaseline(from: snapshot)
      return .countersReset
    }

    var categories = CategorizedTrafficBytes.zero
    var connectionDeltas: [ConnectionTrafficDelta] = []

    for connection in snapshot.connections {
      let delta: TrafficBytes
      if let previous = previousConnections[connection.id] {
        guard
          let existingDelta = TrafficBytes.nonnegativeDelta(
            current: connection.bytes,
            previous: previous
          )
        else {
          continue
        }
        delta = existingDelta
      } else {
        delta = connection.bytes
      }

      let classification = classifier.classify(chains: connection.chains)
      guard classification.category != .unknown else {
        continue
      }

      categories = categories.adding(delta, to: classification.category)
      connectionDeltas.append(
        ConnectionTrafficDelta(
          id: connection.id,
          category: classification.category,
          bytes: delta,
          cumulativeBytes: connection.bytes,
          metadata: connection.metadata,
          startedAt: connection.startedAt
        )
      )
    }

    let unknown = TrafficBytes.residual(
      total: kernelDelta,
      subtracting: categories.classified
    )
    categories = categories.adding(unknown, to: .unknown)
    establishBaseline(from: snapshot)

    return .delta(
      ConnectionDeltaBatch(
        traffic: TrafficDeltaReport(kernel: kernelDelta, categories: categories),
        connections: connectionDeltas
      )
    )
  }

  mutating func reset() {
    previousKernelTotal = nil
    previousConnections = [:]
  }

  private mutating func establishBaseline(from snapshot: ConnectionTrafficSnapshot) {
    previousKernelTotal = snapshot.kernelTotal
    previousConnections = snapshot.connections.reduce(into: [:]) {
      $0[$1.id] = $1.bytes
    }
  }
}

enum ConnectionDeltaResult: Equatable, Sendable {
  case baselineEstablished
  case delta(ConnectionDeltaBatch)
  case countersReset
}

File: platform/windows/MihomoMeter.Windows.Core/Domain/ConnectionDeltaTracker.cs
namespace MihomoMeter.Windows.Core.Domain;

public sealed class ConnectionDeltaTracker
{
    private TrafficBytes? _previousKernelTotal;
    private Dictionary<string, TrafficBytes> _previousConnections = new(StringComparer.Ordinal);

    public ConnectionDeltaResult Consume(
        ConnectionTrafficSnapshot snapshot,
        ProxyClassifier classifier)
    {
        if (_previousKernelTotal is null)
        {
            EstablishBaseline(snapshot);
            return new ConnectionDeltaResult(ConnectionDeltaStatus.BaselineEstablished);
        }

        var kernelDelta = TrafficBytes.NonnegativeDelta(
            snapshot.KernelTotal,
            _previousKernelTotal.Value);
        if (kernelDelta is null)
        {
            EstablishBaseline(snapshot);
            return new ConnectionDeltaResult(ConnectionDeltaStatus.CountersReset);
        }

        var categories = CategorizedTrafficBytes.Zero;
        var connectionDeltas = new List<ConnectionTrafficDelta>();

        foreach (var connection in snapshot.Connections)
        {
            TrafficBytes? delta = _previousConnections.TryGetValue(connection.Id, out var previous)
                ? TrafficBytes.NonnegativeDelta(connection.Bytes, previous)
                : connection.Bytes;
            if (delta is null)
            {
                continue;
            }

            var classification = classifier.Classify(connection.Chains);
            if (classification.Category == TrafficCategory.Unknown)
            {
                continue;
            }

            categories = categories.Adding(delta.Value, classification.Category);
            connectionDeltas.Add(new ConnectionTrafficDelta(
                connection.Id,
                classification.Category,
                delta.Value,
                connection.Bytes,
                connection.Metadata,
                connection.StartedAt));
        }

        var unknown = TrafficBytes.Residual(kernelDelta.Value, categories.Classified);
        categories = categories.Adding(unknown, TrafficCategory.Unknown);
        EstablishBaseline(snapshot);

        return new ConnectionDeltaResult(
            ConnectionDeltaStatus.Delta,
            new ConnectionDeltaBatch(
                new TrafficDeltaReport(kernelDelta.Value, categories),
                connectionDeltas));
    }

    public void Reset()
    {
        _previousKernelTotal = null;
        _previousConnections.Clear();
    }

    private void EstablishBaseline(ConnectionTrafficSnapshot snapshot)
    {
        _previousKernelTotal = snapshot.KernelTotal;
        _previousConnections = snapshot.Connections.ToDictionary(
            connection => connection.Id,
            connection => connection.Bytes,
            StringComparer.Ordinal);
    }
}
